tools/bin/mbedtls-rewrite-branch-style

```diff
@@ -0,0 +1,591 @@
+            new_name_target = self.run_git(['rev-parse', new_name],
+                                           stderr=subprocess.DEVNULL)
+            if old_name_target == new_name_target:
+                self.info('Backup branch {} already points to {}', new_name, ref)
+                return
+        except subprocess.CalledProcessError:
+            # Commonly, new_name does not exist, and that's a good case for us.
+            # If some other error occurred (e.g. ref is invalid), ignore it:
+            # let 'git branch' fail with a more informative error.
+            self.info('Creating backup branch {} for {}', new_name, ref)
+            self.run_git(['branch', '--no-track', new_name, ref])
+
+    def do_rewrite(self, branch: str) -> None:
+        """Rewrite the given branch.
+
+        Do not call this method directly: call rewrite().
+
+        Override this method in child classes to do the desired rewriting.
+
+        This method is called with the current directory in a temporary
+        Git worktree with branch checked out in detached mode.
+        """
+        raise NotImplementedError
+
+    def rewrite(self, branch: str, update_branch: bool = False) -> None:
+        """Rewrite the given branch.
+
+        This method creates a temporary Git worktree and changes to it.
+
+        If update_branch is true and the rewrite succeeds, update branch
+        to the rewritten commit and check it out in the current directory
+        if it was already checked out before.
+        """
+        assert self.original_worktree_path is not None
+        assert self.worktree_path is None
+        update_original_worktree = False
+        original_branch_sha = self.run_git(['rev-parse', branch]).strip()
+        if update_branch and self.checked_out_branch() == branch:
+            if not self.working_directory_is_clean():
+                self.log_error('Refusing to update branch and working tree because the working tree is dirty.')
+                self.log_error('Hint: commit your changes first.')
+                return
+            self.info('Will update original worktree on success')
+            update_original_worktree = True
+        self.worktree_path = self.generate_worktree_path(branch)
+        self.info('Creating worktree: {}', self.worktree_path)
+        self.run_git(['worktree', 'add', '--detach', self.worktree_path])
+        os.chdir(self.worktree_path)
+        self.run_git_checkout(branch)
+        if self.working_branch_name:
+            self.info('Working on new branch {}', self.working_branch_name)
+            self.run_git(['checkout', '--quiet', '-b', self.working_branch_name])
+        self.do_rewrite(branch)
+        self.rewritten_sha = self.head_sha()
+        if update_branch:
+            self.info('Moving branch {} from {} to {}',
+                      branch, original_branch_sha, self.rewritten_sha)
+            self.run_git(['branch', '--force', branch, self.rewritten_sha])
```

Comment:
TODO: This doesn't work if `branch` was checked-out by the user before starting.

Reply:
Indeed.

I'd like to simplify the script's handling of branches. It has too many options — they kind of grew organically based on my convenience when developing and testing, but I don't think they make a nice user interface.

I'm torn between:

* really wanting to make a backup before changing what a branch points to (yes, there's the reflog, but a gc can take it away and who can find the needle in a haystack-sized reflog?).
* not being keen on imposing a name for the backup branch (some people are picky about their branch names).
* not being keen on having so many options to set branch names.

Reply:
This line is failing for me in the update test as well, and I haven't yet figured out how to fix it.

Reply:
To update the branch and a worktree with the branch checked out, I guess the right command would be `git reset --hard $NEW_SHA` in that worktree.

I can't think of why `git branch --force $BRANCH $NEW_SHA` would fail if the branch isn't checked out somewhere.

That leaves the case when the branch is checked out but not in the original worktree. Would it make sense to detect that and update whichever worktree has the branch checked out? Or is that best treated as a user error? If a user runs `mbedtls-rewrite-branch-style --whatever-option-means-to-update-an-existing-branch-name=mybranch` and `mybranch` is checked out in some worktree but they've forgotten about it, should we update that worktree or not? At least we definitely should not update the worktree if it's dirty; should we just update the branch, or refuse to do so?

Reply:
> That leaves the case when the branch is checked out but not in the original worktree. Would it make sense to detect that and update whichever worktree has the branch checked out? Or is that best treated as a user error?

I vote user error. Document somewhere that we don't support it, and just try not to mess everything up if it happens anyway, but I don't think we need to support every single weird corner case that can easily be avoided by users.